esphome/docker-compose.yml

```diff
@@ -0,0 +1,21 @@
+version: "3.7"
+
+services:
+  app_proxy:
+    environment:
+      APP_HOST: $APP_ESPHOME_IP
+      APP_PORT: 6052
+
+  server:
+    image: esphome/esphome:2022.6.2@sha256:1fac6334b85e76c7fd936036891b7a7ee761ee73fd119662b2cb564c3faf818f
+    volumes:
+      - ${APP_DATA_DIR}/data:/config
+      # Needed for flashing new ESPs
+      - /dev:/dev
```

Comment:
Passing all of `/dev` directly through to the container makes me feel a little uneasy. We just need access to `/dev/ttyUSB0` right? Maybe we can lock it down to:

```
devices:
  - "/dev/ttyUSB0:/dev/ttyUSB0"
```

Alternatively, do we need the flashing functionality of ESPHome to work directly on Umbrel hardware? Umbrel is normally run on a server hidden away somewhere and you'd access the web ui on a laptop or something. So having to USB connect a device to your server to flash it through a web UI on your laptop seems like an anti pattern.

If you could plug the device into your laptop and flash it via WebUSB or something it seems like that would be a more logical approach. Thoughts?

Reply:
Ok so just testing the app out looks like it does support this behaviour by default although WebUSB access is disabled due to no SSL. However it sends the user to https://web.esphome.io/?dashboard_wizard which then does allow flashing a bootstrap esphome firmware via WebUSB (since this page has SSL) which will allow the device to be discoverable on the network and the setup can then be continued via the local ESPHome install.

I don't have an ESP32 device to test this flow on but it seems like the best option to me. @ShonP40 do you agree with this? If we can remove the `/dev` access and we can go with this bootstrap option I think this is good to be merged.

If you have an ESP32 device you could test this flow with it would be great if you could confirm.

Reply:
One thing that definitely would be worth testing is that after flashing the bootstrap firmware the Umbrel can detect the device with the current compose config. Often network discovery stuff is done via UDP scanning which I don't think is possible from a container unless you add `network_mode: "host"`.

Reply:
I can limit it to `ttyUSB0` but then it'll only detect one ESP at a time.

And about discovery, you are right, it doesn't detect ESPs that are already on the network and doesn't update the online indicator.
However, this only breaks the auto adopt functionality, all the other functions continue to work normally.

So it's up to you guys if we should make this run on the host network or not.

Reply:
If you do want auto adopt to work (thus requiring this container to run on the host network), I'm guessing that users will want Home Assistant on the host network as well (so it'll be able to detect these ESPs automatically)

Reply:
>And about discovery, you are right, it doesn't detect ESPs that are already on the network and doesn't update the online indicator.
However, this only breaks the auto adopt functionality, all the other functions continue to work normally.

So does this mean you can for example update the firmware once the device is setup? Or are you saying it's not possible to actually complete the setup process?

Basically I'm just trying to figure out (without having an ESP32 device to test) if this is in a usable state without host networking and hardware access or not. If it is, I think we should ship it like that, then maybe revisit adding more permissions if there are significant requests from users for it.

>If you do want auto adopt to work (thus requiring this container to run on the host network), I'm guessing that users will want Home Assistant on the host network as well (so it'll be able to detect these ESPs automatically)

As a general rule we don't want to allow host networking for apps since it also violates the app sandbox. There is a workaround for mDNS where we just run some sort of DNS proxy on the host that resolves mDNS requests to local IPs and forwards on normal DNS requests. Then configure app containers to use this for DNS. That should allow apps like ESPHome and Home Assistant to resolve .local domain names without requiring host networking access. We don't have this implemented yet however.

We could potentially ship host networking as a short term workaround until that's implemented but it causes a lot of other problems. We have an authentication proxy sitting in front of all Umbrel app web UIs that verify the user is logged in to the Umbrel before proxying traffic through. This breaks if the app is directly listening on host ports. It also means we don't get explicit control over which ports are being opened on the host to prevent port collisions between apps.

In your opinion is ESPHome useful without host networking or hardware access? If so I think we should ship like that with the goal of enabling mDNS resolution in the future at the Umbrel level without having to give apps host networking. So the app will just suddenly be even better in the future once we solve the mDNS problem and the auto discovery logic will start working.

Reply:
Yea, ESPHome is completely useful without host networking.
You can update devices OTA.

The only caveat would be that you'll need to make it upload using the device's IP instead of its .local domain by adding
`use_address: ESP IP` to your ESPHome config (for each device)

Reply:
Ok sweet, that's the same way Home Assistant works atm, you need to manually enter an IP to pair a device. Lets ship like this and we'll try to get a good solution for mDNS discovery implemented in Umbrel.

Thanks for your help @ShonP40!